File: C#/CountingSort.cs
using System;

class CountingSort
{
    public static void Sort(int[] arr)
    {
        int max = arr[0], min = arr[0];
        foreach (var num in arr)
        {
            if (num > max) max = num;
            if (num < min) min = num;
        }

        int range = max - min + 1;
        int[] count = new int[range];
        int[] output = new int[arr.Length];

        foreach (var num in arr)
            count[num - min]++;

        for (int i = 1; i < range; i++)
            count[i] += count[i - 1];

        for (int i = arr.Length - 1; i >= 0; i--)
        {
            output[count[arr[i] - min] - 1] = arr[i];
            count[arr[i] - min]--;
        }

        for (int i = 0; i < arr.Length; i++)
            arr[i] = output[i];
    }

    public static void Main()
    {
        int[] arr = { 4, 2, 2, 8, 3, 3, 1 };
        Sort(arr);
        Console.WriteLine(string.Join(" ", arr));
    }
}

File: Kotlin/countingSort.kt
fun countingSort(arr: IntArray): IntArray {
    if (arr.isEmpty()) return arr

    val max = arr.maxOrNull() ?: return arr
    val min = arr.minOrNull() ?: return arr
    val range = max - min + 1
    val count = IntArray(range)
    val output = IntArray(arr.size)

    for (num in arr) {
        count[num - min]++
    }

    for (i in 1 until count.size) {
        count[i] += count[i - 1]
    }

    for (i in arr.indices.reversed()) {
        output[count[arr[i] - min] - 1] = arr[i]
        count[arr[i] - min]--
    }

    return output
}

fun main() {
    val arr = intArrayOf(4, 2, 2, 8, 3, 3, 1, 0, -8)
    val sortedArr = countingSort(arr)
    println(sortedArr.joinToString(", "))
}